### ISSUE TITLE
docs: Review documentation site generation for versioned OPA docs

### ISSUE THREAD

**Issue by charlieegan3** [open-policy-agent/opa#7037]
Currently we use Hugo to build a website for OPA's homepage and documentation content. There are some sharp edges in this setup:

- developers must use a compatible Hugo version, 0.113.0. It's not currently possible for us to upgrade without updating the old content due to the use of a front matter field.
- there are two content dirs, one containing the regular Hugo content under website and another containing the current docs content. Running generation scripts is required to be able to build a local version of the site for testing docs changes.
- Versioned docs are generated by copying the docs content from all versions into a generated content dir under docs. This generates paths the like `/docs/v0.67.1/`. The challenge that this poses is that content from past versions of OPA must work with the current Hugo site and configuration. This is something which, if broken, is only spotted after merging to main when all the contents is updated.

These issues make it hard to maintain the docs, and, I feel, place the bar to drive by docs contributions too high.

I think we should consider a system where:

- all content from all versions is easily available when developing the site, rather than needing to run lots of generation scripts and pull state from git.
- Use a static site generator which has native versioned docs support.
- Review the content where versioning is important, the note from https://docusaurus.io/docs/versioning is interesting:

> Most of the time, you don't need versioning as it will just increase your build time, and introduce complexity to your codebase. Versioning is best suited for websites with high-traffic and rapid changes to documentation between versions. If your documentation rarely changes, don't add versioning to your documentation.

What if we only versioned a limited set of pages, for example:

- https://www.openpolicyagent.org/docs/latest/policy-reference/
- https://www.openpolicyagent.org/docs/latest/rest-api/
- https://www.openpolicyagent.org/docs/latest/cli/

**Comment by srenatus** [comment]
Sorry you ran into this problem: the website build only doing the latest version was intentionally done for dev and preview (https://github.com/open-policy-agent/opa/issues/4379) because it took so long to always build everything. As a workaround, when doing "global changes" that need the previous versions, I've tended [to revert this Makefile change](https://github.com/open-policy-agent/opa/pull/4474/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R453), which would make netlify _build everything_ for a preview.

Taking a step back as suggested, I think the versioning isn't needed, or it's even harmful. Besides complicating the build process (since Hugo wasn't made for this), it also makes people find old content, use old images, etc.

Where it *is important* is when it comes to builtins and when they've been introduced -- but we already have introduced those version tags: 
<img width="492" alt="image" src="https://github.com/user-attachments/assets/9e5e7047-857b-4546-9db2-ad10314a3346">

> What if we only versioned a limited set of pages, for example:
> 
> * https://www.openpolicyagent.org/docs/latest/policy-reference/
> * https://www.openpolicyagent.org/docs/latest/rest-api/
> * https://www.openpolicyagent.org/docs/latest/cli/

What if we didn't? 😅 The policy reference hasn't changed much since "every". There's going to be a v1 release eventually and we will stop pretending that "in", "every", "contains" are optional future stuff -- it's just Rego.

The REST API hasn't changed at all, I think, in the last years.

The CLI changes are usually minimal, but it would help if you had a chance to figure out when a flag was introduced -- but really, I think the scenario is like this: 
1. you find a flag in some docs
2. you try it locally, it's not found
3. you try to figure out if it's in the latest version

Right now, there's no easy way to figure out _when_ a flag was introduced. You'll have to flip through all the versioned docs...

tl;dr: I think our current approach to versioning docs is rather blunt and worth revisiting. I've never referred to anything but the "latest" docs (and maybe "edge"), and wouldn't shed a single tear if they went away.

**Comment by charlieegan3** [comment]
> What if we didn't?

I mean, the thought had crossed my mind! 😅 

I suppose the other way of thinking about it is that OPA is a more simple component to deploy and update than say, Kubernetes. Kubernetes have a year support period for releases: https://kubernetes.io/releases/patch-releases/#support-period, and various managed offerings have a range of active versions exposed to users via channels.

In OPA we don't really operate releases in the same way and so even though the usage of dated OPA versions is high, it's less important that the docs are timecapsuled for every version for the reasons that Stephan outlines above.

**Comment by charlieegan3** [comment]
Thank you stale bot, we'll take it from here!

I am starting some work on this one to build a browsable docs archive similar to what Kubernetes and Istio do with their [docs](https://v1-29.docs.kubernetes.io/releases/) [archives](https://istio.io/archive/).

The benefit of doing this is that we decouple the versioned content from the current implementation of the site. This means that:

- We can improve the presentation of the site without needing to babysit the old content in the same wrapper.
- We can migrate to other docs tooling that will be easier to use.
- We can reduce the time taken to generate the docs.
- We can make use of some prior art for in-browser, live code evaluation.

High level plan of operations here:

- Build versioned deployments similar to `https://v1-29.docs.kubernetes.io`, including automations.
- Update the version switcher to point to these instead of the path based versions, & have appropriate redirects in place too.
- Start to build a new OPA docs presentation and switch to this going forward.

**Comment by charlieegan3** [comment]
Some progress to report here, is that versioned docs sites are now linked from the production docs nav. This means that we have decoupled the old versions of the site from main, and thus can start moving forward with a new version of the site.

**Comment by charlieegan3** [comment]
https://github.com/open-policy-agent/opa/pull/7531 will ensure that any old links pointing to `/docs/vx.y.z/foo/bar` pages will be redirected correctly. These are not linked to in the site, but they might be linked from private docs and elsewhere online.

**Comment by charlieegan3** [comment]
With #7534 merged, we are in a good position to release the new site this week.
